include/xgboost/data.h

```diff
@@ -270,20 +277,22 @@ class SparsePage {
   size_t Size() { return offset.Size() - 1; }
 };
 
+template<typename T>
 class BatchIteratorImpl {
  public:
   virtual ~BatchIteratorImpl() {}
   virtual BatchIteratorImpl* Clone() = 0;
```

Comment:
do we need these clones? multiple batch iterators could potentially share the same impl and safely destruct.

Reply:
I wanted to stick with pure refactoring for this PR. You are welcome to send another one to get rid of the clones.